Fix Connection OpenTry relaying from Sovereign to Cosmos

soareschen opened the pull request:
Closes: #332

soareschen commented:
Currently the test fails with the error:

```
Error: Status { code: Unknown, message: "failed to execute message; message index: 1: connection handshake open try failed: failed connection state verification for client (08-wasm-0): IBC validation/execution context error: ICS02 Client error: other error: `Keys do not match. Key in proof: KeyHash(\"a99b032d28ad20144acec712a9b857443176fda6ca1f13084e700c6263e5b303\"). Expected key: KeyHash(\"ad7027e03ff33393de537629edb3adea6133c8c529e3382a2b1494b35428fa73\").`: wasm contract call failed [cosmos/ibc-go/modules/light-clients/08-wasm/types/vm.go:156] With gas wanted: '18446744073709551615' and gas used: '168914' ", metadata: MetadataMap { headers: {"content-type": "application/grpc", "x-cosmos-block-height": "31"} }, source: None }
```

rnbguy commented:
Currently, the proof verification fails, as the proof fetched at height `h` generates the root hash for height `h-1`.

```console
rollup user hash at RollupHeight { slot_number: 12 }: [97, 167, 3, 147, 202, 241, 94, 66, 28, 61, 41, 218, 61, 69, 213, 104, 202, 238, 108, 172, 246, 92, 224, 101, 180, 29, 53, 5, 15, 5, 113, 171]
...
update client payload: rollup user hash at RollupHeight { slot_number: 13 }: [2, 242, 237, 13, 231, 164, 245, 166, 156, 165, 41, 45, 47, 255, 90, 111, 62, 154, 193, 31, 222, 104, 193, 204, 207, 200, 58, 187, 200, 120, 246, 11]
Error: Status { code: Unknown, message: "failed to execute message; message index: 1: connection handshake open try failed: failed connection state verification for client (08-wasm-0): IBC validation/execution context error: ICS02 Client error: other error: `Root hashes do not match. Actual root hash: [97, 167, 3, 147, 202, 241, 94, 66, 28, 61, 41, 218, 61, 69, 213, 104, 202, 238, 108, 172, 246, 92, 224, 101, 180, 29, 53, 5, 15, 5, 113, 171]. Expected root hash: [2, 242, 237, 13, 231, 164, 245, 166, 156, 165, 41, 45, 47, 255, 90, 111, 62, 154, 193, 31, 222, 104, 193, 204, 207, 200, 58, 187, 200, 120, 246, 11].`: wasm contract call failed [cosmos/ibc-go/modules/light-clients/08-wasm/types/vm.go:156] With gas wanted: '18446744073709551615' and gas used: '160517' ", metadata: MetadataMap { headers: {"content-type": "application/grpc", "x-cosmos-block-height": "35"} }, source: None }
```

I added a comment on the code for this.

rnbguy commented:
Sovereign integration test passes now.

I added `sleep`s before the sovereign rollup queries.
- So I removed the `slot + 1` for `leader_getHead`.
- Later I also found that, the proof queries need to be queried at `slot + 2`.

We need to confirm this with the sovereign team.

soareschen commented on crates/sovereign/sovereign-chain-components/src/sovereign/impls/sovereign_to_cosmos/client/update_client_payload.rs:
```diff
@@ -126,10 +131,22 @@ where
             .await
             .map_err(|e| eyre!("Error creating headers from DA chain: {e:?}"))?;
 
+        let chain_status = chain.query_chain_status_at_height(target_height).await?;
```
I am not sure if we have to have a generalized version of `query_chain_status_at_height`. It seems that the use is specific to Sovereign. So perhaps we can do the JSON RPC query directly.

rnbguy replied:
The way I see, cosmos chain doesn't need it, becuase we call `build_header` from hermes.

This step is basically `build_header` for sovereign. Considering this, I will let you make the decision.

soareschen replied:
I moved the trait to `sovereign-rollup-components`, since it is sovereign-specific.

soareschen commented:
> The ConnectionOpenTry integration test for sovereign to cosmos passes now !
> 
> I added `sleep`s before the sovereign rollup queries - so that the rollup is ready for query.
> 
>     * So I removed the `slot + 1` for `leader_getHead`.
> 
>     * Later I also found that, the proof queries need to be queried at `slot + 2`.

Instead of sleeping and using slot + 2, I make it such that `query_chain_status` returns `slot_number - 2`, and it somehow works.
